### PR TITLE
Session & EthereumPeer refactoring & unit tests for EthereumPeer

### PR DESCRIPTION
**Pull request by gumb0**
Refactoring trying to achieve better testability of EthereumPeer class.
Currently work in progress.

This PR is based on #3466 and first commit is from there.

### CODE REVIEW

File: libp2p/Session.h [hunk 1]
```diff
@@ -158,11 +193,25 @@ class Session: public std::enable_shared_from_this<Session>
...
 	std::shared_ptr<Framing> getFraming(uint16_t _protocolID);
 	void multiplexAll();
 };
 
+template <class PeerCap>
+std::shared_ptr<PeerCap> capabilityFromSession(SessionFace const& _session, u256 const& _version = PeerCap::version())
+{ 
+	try 
+	{ 
+		return std::static_pointer_cast<PeerCap>(_session.capabilities().at(std::make_pair(PeerCap::name(), _version)));
```
Comment (chriseth) [inline libp2p/Session.h]: This is only used in networking code, right? So we could spend the extra CPU cycles and use `dynamic_pointer_cast` for added safety. What do you think?

Comment (gumb0) [inline libp2p/Session.h]: That would basically check that our initialization code correctly registers Capability classes with correct names. Maybe some debug-only check would be enough? Like adding `assert(dynamic_pointer_cast<PeerCap>(cap))`

The result value returned from this function is almost never checked for nullptr anyway

File: libethereum/EthereumHost.cpp [hunk 2]
```diff
@@ -49,14 +49,278 @@ const char* EthereumHostTrace::name() { return EthPurple "^" EthGray "  "; }
...
+private:
+	EthereumHost& m_host;
+};
+
+class EthereumHostData: public EthereumHostDataFace
+{
+public:
+	EthereumHostData(BlockChain const& _chain, OverlayDB const& _db): m_chain(_chain), m_db(_db) {}
+
+	pair<bytes, unsigned> blockHeaders(RLP const& _blockId, unsigned _maxHeaders, u256 _skip, bool _reverse) const override
```
Comment (gumb0) [inline libethereum/EthereumHost.cpp]: This code is moved here from `EthereumPeer::interpret()`

**Comment by gumb0** [comment]
Rebased on latest develop

**Pull request merged**